client-tools/Shell/TelemetricsHandler.cpp

```diff
@@ -0,0 +1,348 @@
+      return url.substr(0, p - url.c_str());
+    } else if (*p == '/') {
+      if (++slashes == 3) {
+        return url.substr(0, p - url.c_str());
+      }
+    }
+    ++p;
+  }
+
+  return url;
+}
+}  // namespace
+
+namespace arangodb {
+TelemetricsHandler::TelemetricsHandler(ArangoshServer& server,
+                                       bool sendToEndpoint)
+    : _server(server), _sendToEndpoint(sendToEndpoint) {}
+
+TelemetricsHandler::~TelemetricsHandler() {
+  if (_telemetricsThread.joinable()) {
+    _telemetricsThread.join();
+  }
+}
+
+Result TelemetricsHandler::checkHttpResponse(
+    std::unique_ptr<httpclient::SimpleHttpResult> const& response) {
+  using basics::StringUtils::concatT;
+  using basics::StringUtils::itoa;
+  if (response == nullptr || !response->isComplete()) {
+    return {TRI_ERROR_INTERNAL, this->_httpClient->getErrorMessage()};
+  }
+  if (response->wasHttpError()) {
+    auto errorNum = TRI_ERROR_INTERNAL;
+    std::string errorMsg = response->getHttpReturnMessage();
+    std::shared_ptr<velocypack::Builder> bodyBuilder;
+    // Handle case of no body:
+    try {
+      bodyBuilder = response->getBodyVelocyPack();
+      velocypack::Slice error = bodyBuilder->slice();
+      if (!error.isNone() && error.hasKey(StaticStrings::ErrorMessage)) {
+        errorNum = ErrorCode{
+            error.get(StaticStrings::ErrorNum).getNumericValue<int>()};
+        errorMsg = error.get(StaticStrings::ErrorMessage).copyString();
+      }
+    } catch (...) {
+    }
+    return {errorNum,
+            concatT("got invalid response from server: HTTP ",
+                    itoa(response->getHttpReturnCode()), ": ", errorMsg)};
+  }
+
+  return {TRI_ERROR_NO_ERROR};
+}
+
+void TelemetricsHandler::fetchTelemetricsFromServer() {
+  _telemetricsFetchedInfo.clear();
+  std::string const url = "/_admin/telemetrics";
+  Result result = {TRI_ERROR_NO_ERROR};
+  uint32_t timeoutInSecs = 1;
+  while (!_server.isStopping()) {
+    ClientManager clientManager(
+        _server.getFeature<HttpEndpointProvider, ClientFeature>(),
+        Logger::FIXME);
+    std::unique_lock lk(_mtx);
+
+    _httpClient = clientManager.getConnectedClient(true, false, false, 0);
+    if (_httpClient != nullptr && _httpClient->isConnected()) {
+      auto& clientParams = _httpClient->params();
+      clientParams.setRequestTimeout(30);
+      lk.unlock();
+      std::unique_ptr<httpclient::SimpleHttpResult> response(
+          _httpClient->request(rest::RequestType::GET, url, nullptr, 0));
+      lk.lock();
+
+      result = this->checkHttpResponse(response);
+      if (auto errorNum = result.errorNumber();
+          errorNum == TRI_ERROR_NO_ERROR ||
+          errorNum == TRI_ERROR_HTTP_FORBIDDEN) {
+        _telemetricsFetchedInfo.add(response->getBodyVelocyPack()->slice());
+        _telemetricsFetchResponse = std::move(result);
+        break;
+      } else {
+        _telemetricsFetchResponse = std::move(result);
+      }
```

Comment:
`_telemetricsFetchResponse` is set in both the `if` and `else` cases, so the code can be simplified to:
```suggestion
      _telemetricsFetchResponse = std::move(result);
      if (auto errorNum = result.errorNumber();
          errorNum == TRI_ERROR_NO_ERROR ||
          errorNum == TRI_ERROR_HTTP_FORBIDDEN) {
        _telemetricsFetchedInfo.add(response->getBodyVelocyPack()->slice());
        break;
      }
```